src/multihasher.rs

```diff
@@ -1,25 +1,78 @@
 use std::collections::VecDeque;
 use std::fmt;
 
+use async_trait::async_trait;
 use libp2p_core::multihash::Multihash;
 use multihash_codetable::MultihashDigest;
 
 use crate::utils::convert_multihash;
 
+/// Errors that can be produced by [`Multihasher`] trait.
+#[derive(Debug, thiserror::Error)]
+pub enum MultihasherError {
+    /// [`Multihasher`] can not handle the specified multihash code.
+    #[error("Unknown multihash code")]
+    UnknownMultihashCode,
+
+    /// Bigger [`Multihash`] is needed.
+    ///
+    /// Size of [`Multihash`] is specified as generic const when [`Behaviour`]
+    /// is constructed.
+    ///
+    /// [`Behaviour`]: crate::Behaviour
+    #[error("Invalid multihash size")]
+    InvalidMultihashSize,
+
+    /// Custom error.
+    ///
+    /// This error will be logged and the specified `input` will be ignored.
+    #[error("{0}")]
+    Custom(String),
+
+    /// Custom error that causes [`Stream`] to close.
+    ///
+    /// This error will be logged and the [`Stream`] which delivered the
+    /// `input` will be closed.
+    ///
+    /// [`Stream`]: libp2p_swarm::Stream
+    #[error("FATAL: {0}")]
```

Comment:
```suggestion
    #[error("Hashing failed: {0}")]
```

I don't like screaming nature in logs :d

Reply:
But `Custom` is also a failure, but without closing the stream. What about `Hashing failure' for `Custom` and `Fatal hashing failure` for `CustomFatal`?